client/x/evmstaking/keeper/abci.go

```diff
@@ -29,96 +21,28 @@ func (k *Keeper) EndBlock(ctx context.Context) (abci.ValidatorUpdates, error) {
 	if err != nil {
 		return nil, err
 	}
 
-	// make an array with each entry being the validator address, delegator address, and the amount
-	var unbondedEntries []UnbondedEntry
-
-	for _, dvPair := range matureUnbonds {
-		validatorAddr, err := k.validatorAddressCodec.StringToBytes(dvPair.ValidatorAddress)
-		if err != nil {
-			return nil, errors.Wrap(err, "validator address from bech32")
-		}
-
-		delegatorAddr, err := k.authKeeper.AddressCodec().StringToBytes(dvPair.DelegatorAddress)
-		if err != nil {
-			return nil, errors.Wrap(err, "delegator address from bech32")
-		}
-
-		ubd, err := (k.stakingKeeper).GetUnbondingDelegation(ctx, delegatorAddr, validatorAddr)
-		if err != nil {
-			return nil, err
-		}
-
-		// TODO: parameterized bondDenom
-		bondDenom := sdk.DefaultBondDenom
-
-		// loop through all the entries and process unbonding mature entries
-		for i := range len(ubd.Entries) {
-			entry := ubd.Entries[i]
-			if entry.IsMature(ctxTime) && !entry.OnHold() {
-				// track undelegation only when remaining or truncated shares are non-zero
-				if !entry.Balance.IsZero() {
-					amt := sdk.NewCoin(bondDenom, entry.Balance)
-					// TODO: check if it's possible to add a double entry in the unbondedEntries array
-					unbondedEntries = append(unbondedEntries, UnbondedEntry{
-						validatorAddress: dvPair.ValidatorAddress,
-						delegatorAddress: dvPair.DelegatorAddress,
-						amount:           amt.Amount,
-					})
-				}
-			}
-		}
-	}
-
-	valUpdates, err := k.stakingKeeper.EndBlocker(ctx)
-	if err != nil {
-		return nil, err
-	}
+	log.Debug(ctx, "Processing mature unbonding delegations", "count", len(unbondedEntries))
 
 	for _, entry := range unbondedEntries {
-		delegatorAddr, err := k.authKeeper.AddressCodec().StringToBytes(entry.delegatorAddress)
+		delegatorAddr, err := k.authKeeper.AddressCodec().StringToBytes(entry.DelegatorAddress)
 		if err != nil {
 			return nil, errors.Wrap(err, "delegator address from bech32")
 		}
 
-		spendableAmount := k.bankKeeper.SpendableCoin(ctx, delegatorAddr, sdk.DefaultBondDenom).Amount
-		if spendableAmount.IsZero() {
-			log.Warn(ctx, "No spendable coins for undelegation",
-				errors.New("no spendable coins for undelegation"),
-				"delegator", entry.delegatorAddress,
-				"validator", entry.validatorAddress,
-				"original_amount", entry.amount.String())
-
-			continue
-		}
-
-		// If the requested undelegation amount is greater than the spendable amount, set the real undelegation amount to
-		// the total spendable amount.
-		if entry.amount.GT(spendableAmount) {
-			entry.amount = spendableAmount
-			log.Warn(ctx, "Spendable amount is less than the requested undelegation amount",
-				errors.New("spendable amount is less than the requested undelegation amount"),
-				"delegator", entry.delegatorAddress,
-				"validator", entry.validatorAddress,
-				"requested_amount", entry.amount.String(),
-				"spendable_amount", spendableAmount.String())
-		}
```

Comment:
Should we add tests to ensure this invariance is not broken? cc @limengformal 

We should also add the crisis module for invariance checks, cc @edisonz0718.

Reply:
Will add a test to cover that this invariance is not broken in the following PR.